### PR TITLE
bazel: monitoring docs gen

### PR DESCRIPTION
**Pull request by burmudar**
### Changes
improve `write_generated_to_source_files` to:
- allow prefixes to be stripped from generated files
- specify a different destination
- handle being called from a different package than being next to the src target

This seemed like an easy change on the outset but it turned out to be quite involved. Some of the problems I encountered while doing this which required the above changes:

* monitoring-generator copies to a total different path in the repo route than what it normally outputs - ie. docs/alerts.md, docs/dashboard.md -> doc/admin/observability/
* `//monitoring:generate_config` outputs are local to it and AFAIK you can copy/move files to the parent package/dir
* The `//monitoring:generate_config` already does 90% of what we needed, but the outputs needed to be massaged a bit to fit out use case.
* When one target calls another target in a macro and they're not in the same package, the "outside" targets outputs will have it's package name prepended. Which is why we use `root_paths` during copy to have the copied file paths the same as if the target was called in the same package
Example:
```
bazel run //:write_monitoring_docs
-> calls //monitoring:generate_config
-> results in outputs being available under `bazel-out/.../monitoring/outputs/`

bazel run //monitoring:write_monitoring_docs
-> calls //monitoring:generate_config
-> results in outputs  being available under `bazel-out/.../outputs/
```

## Test plan
run locally

- [ ] Other generated targets still work

### CODE REVIEW

File: dev/write_generated_to_source_files.bzl [hunk 1]
```diff
@@ -1,19 +1,45 @@
 load("@aspect_bazel_lib//lib:directory_path.bzl", "make_directory_path")
 load("@aspect_bazel_lib//lib:copy_to_directory.bzl", "copy_to_directory")
 load("@aspect_bazel_lib//lib:write_source_files.bzl", "write_source_files")
+load("@bazel_skylib//lib:paths.bzl", "paths")
+
+def write_generated_to_source_files(name, src, files, dest = "", strip_prefix = "", verbose_copy=True, **kwargs):
+
```
Comment (burmudar) [inline dev/write_generated_to_source_files.bzl:7]:
```suggestion
```

File: monitoring/main.go [hunk 2]
```diff
@@ -1,5 +1,4 @@
-//go:generate go build -o /tmp/monitoring-generator
-//go:generate /tmp/monitoring-generator
+//go:generate bazel run //:write_monitoring_docs
```
Comment (jhchabran) [inline monitoring/main.go:1]: This is still a `go:generate` statement, even if it's calling Bazel. This should instead be called by `//dev:write_all`.

File: BUILD.bazel [hunk 3]
```diff
@@ -329,7 +330,22 @@ exports_files([
...
 ])
 
-# stamp_tags(
-#     name = "tags",
-#     remote_tags = ["""($stamp.STABLE_VERSION // "0.0.0")"""],
-# )
+write_generated_to_source_files(
+    name = "write_monitoring_docs",
+    src = "//monitoring:generate_config",
+    dest = "doc/admin/observability/",
```
Comment (jhchabran) [inline BUILD.bazel:336]: This looks fragile, it means that we have a target being declared in the root repo, that does things with a given package and write it in a different place. If we keep that logic, it's going to be really hard to understand what's going to happen when similar rules pops up more.

File: BUILD.bazel [hunk 4]
```diff
@@ -329,7 +330,22 @@ exports_files([
...
+    # - docs describing dashboards and alerts
+    # the grafana dashboards and prometheus config are handled by other targets under //monitoring
+    # but we need to write the docs back to the repo root, and in Bazel we can't use relative paths
+    # to go to the parents directory (i.e one up from monitoring is the repo root)
+    files = [
+        "outputs/docs/alerts.md",
+        "outputs/docs/dashboards.md",
+    ],
+    # since :generate_config stores all the generated files under outputs, we want to copy them out without "outputs/"
+    strip_prefix = "outputs/docs/",
```
Comment (jhchabran) [inline BUILD.bazel:349]: neat 💯

File: dev/write_generated_to_source_files.bzl [hunk 5]
```diff
@@ -1,19 +1,44 @@
...
+    # BUT when we're in different packages we need to replace the src target package root, since
+    # all it's outputs path will have the src package as part of the path. So set root_paths to
+    # the src target package path, which means it will be removed as part of copying
+    if currentTarget.package != srcTarget.package:
+        copy_opts["root_paths"] = [
+            srcTarget.package
+        ]
+        # copy_opts["replace_prefixes"] = {
+        #     srcTarget.package: "",
+        # }
```
Comment (jhchabran) [inline dev/write_generated_to_source_files.bzl:24]: left-over?

**Review by jhchabran — CHANGES_REQUESTED** [review]

**Comment by jhchabran** [comment]
Need to add the target itself to `//dev:write_all_generated`

File: doc/admin/observability/BUILD.bazel [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+
+write_generated_to_source_files(
+    name = "write_monitoring_docs",
+    src = "//monitoring:generate_config",
+    # :generate_config creates an outputs folder with:
+    # - grafana dashboards
+    # - prometheus config
+    # - docs describing dashboards and alerts
+    files = [
+        "monitoring/outputs/docs/alerts.md",
```
Comment (burmudar) [inline doc/admin/observability/BUILD.bazel:22]: I am not a fan of repeating `monitoring/outputs/docs` here and in `strip_prefix` but it is required due to how `//monitoring:generate_config` works atm.

**Review by jhchabran — APPROVED** [review]

**Comment by mrnugget** [comment]
How do I run this? Nothing I try regenerates docs.

**Comment by mrnugget** [comment]
I had to manually run this: `cp bazel-bin/monitoring/outputs/docs/* doc/admin/observability/` Otherwise the docs aren't updated.

**Comment by mrnugget** [comment]
No, doesn't work:

```
$ git diff monitoring
diff --git a/monitoring/definitions/git_server.go b/monitoring/definitions/git_server.go
index 0c4a8d5502..c17c730d2c 100644
--- a/monitoring/definitions/git_server.go
+++ b/monitoring/definitions/git_server.go
@@ -45,7 +45,7 @@ func GitServer() *monitoring.Dashboard {
                                Rows: []monitoring.Row{
                                        {
                                                {
-                                                       Name:        "go_routines",
+                                                       Name:        "fooooooooooooooooooooogo_routines",
                                                        Description: "go routines",
                                                        Query:       "go_goroutines{app=\"gitserver\", instance=~`${shard:regex}`}",
                                                        NoAlert:     true,
$ bazel build //doc/admin/observability:write_monitoring_docs
INFO: Analyzed target //doc/admin/observability:write_monitoring_docs (0 packages loaded, 0 targets configured).
INFO: Found 1 target...
Target //doc/admin/observability:write_monitoring_docs up-to-date:
  bazel-bin/doc/admin/observability/write_monitoring_docs_update.sh
INFO: Elapsed time: 0.113s, Critical Path: 0.00s
INFO: 1 process: 1 internal.
INFO: Build completed successfully, 1 total action


$ git diff doc

```

No doc generated

**Comment by burmudar** [comment]
🤔 I'll take a look!

Thanks for reporting it

**Comment by burmudar** [comment]
> 🤔 I'll take a look!
> 
> Thanks for reporting it

@mrnugget my bad it should've been `bazel run` instead of `bazel build` but regardless it isn't obvious and it should have done it behind the scenes in `sg gen`. I'll update it next week!

```
bazel run //dev:write_all_generated
```
```
-### Git Server: Golang runtime monitoring
+### Git Server Rawr: Golang runtime monitoring

 #### gitserver: go_goroutines

@@ -7437,7 +7437,7 @@ Query: `max by(instance) (go_gc_duration_seconds{job=~".*gitserver"})`

 <br />

-### Git Server: Kubernetes monitoring (only available on Kubernetes)
+### Git Server Rawr: Kubernetes monitoring (only available on Kubernetes)

 #### gitserver: pods_available_percentage

diff --git a/monitoring/definitions/git_server.go b/monitoring/definitions/git_server.go
index 0c4a8d5502..8a4800ae1d 100644
--- a/monitoring/definitions/git_server.go
+++ b/monitoring/definitions/git_server.go
@@ -24,7 +24,7 @@ func GitServer() *monitoring.Dashboard {

        return &monitoring.Dashboard{
                Name:        "gitserver",
-               Title:       "Git Server",
+               Title:       "Git Server Rawr",
                Description: "Stores, manages, and operates Git repositories.",
                Variables: []monitoring.ContainerVariable{
                        {
```

**Pull request merged**